File: DataStructures/Trie/Trie.js
/**
 * @name Node 节点
 * @description 树的节点
 */
class Node {
  constructor(isWord = false) {
    this.isWord = isWord // 是否是单词节点
    this.next = new Map() // 映射
  }
}

/**
 * @name Trie 字典树 前缀数
 */
class Trie {
  constructor() {
    this.root = new Node()
    this.size = 0
  }

  getSize() {
    return this.size
  }

  // 添加单词
  add(word) {
    let cur = this.root

    for (let i = 0; i < word.length; i++) {
      const c = word[i]
      if (cur.next.get(c) === undefined) {
        cur.next.set(c, new Node())
      }
      cur = cur.next.get(c)
    }

    if (!cur.isWord) {
      cur.isWord = true
      this.size++
    }
  }

  // 查询单词
  contains(word) {
    let cur = this.root

    for (let i = 0; i < word.length; i++) {
      const c = word[i]
      if (cur.next.get(c) === undefined) return false
      cur = cur.next.get(c)
    }
    return cur.isWord
  }

  // 是否是前缀
  isPrefix(word) {
    let cur = this.root
    for (let i = 0; i < word.length; i++) {
      const c = word[i]
      if (cur.next.get(c) === undefined) return false
      cur = cur.next.get(c)
    }
    return true
  }
}

module.exports = Trie

File: Java/leetcode/208/Trie.java
// https://leetcode-cn.com/problems/implement-trie-prefix-tree/
// leetcode 208. 实现 Trie (前缀树) pass

import java.util.TreeMap;

/**
 * Trie 字典树、前缀数
 */
public class Trie {
  private class Node {
    public boolean isWord;
    public TreeMap<Character, Node> next;

    public Node(boolean isWord) {
      this.isWord = isWord;
      this.next = new TreeMap<>();
    }

    public Node() {
      this(false);
    }
  }

  private Node root;
  private int size;

  public Trie() {
    root = new Node();
    size = 0;
  }

  public int getSize() {
    return size;
  }

  public void insert(String word) {
    Node cur = root;

    for (int i = 0; i < word.length(); i++) {
      char c = word.charAt(i);

      if (cur.next.get(c) == null) {
        cur.next.put(c, new Node());
      }

      cur = cur.next.get(c);
    }

    if (!cur.isWord) {
      cur.isWord = true;
      size++;
    }
  }

  public boolean search(String word) {
    Node cur = root;

    for (int i = 0; i < word.length(); i++) {
      char c = word.charAt(i);

      if (cur.next.get(c) == null) {
        return false;
      }

      cur = cur.next.get(c);
    }

    return cur.isWord;
  }

  public boolean startsWith(String word) {
    Node cur = root;

    for (int i = 0; i < word.length(); i++) {
      char c = word.charAt(i);

      if (cur.next.get(c) == null) {
        return false;
      }

      cur = cur.next.get(c);
    }

    return true;
  }
}